sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1142 @@
+        target_neighbors = self._select_target_neighbors_wrapper(
+            X_valid, y_valid, classes)
+
+        # Compute the gradient part contributed by the target neighbors
+        grad_static = self._compute_grad_static(X_valid, target_neighbors)
+
+        # Compute the pull loss coefficient
+        pull_loss_coef = (1. - self.weight_push_loss) / self.weight_push_loss
+        grad_static *= pull_loss_coef
+
+        # Decide how to store the impostors
+        if self.impostor_store == 'sparse':
+            use_sparse = True
+        elif self.impostor_store == 'list':
+            use_sparse = False
+        else:
+            # auto: Use a heuristic based on the data set size
+            use_sparse = X_valid.shape[0] > 6500
+
+        # Create a dictionary of parameters to be passed to the optimizer
+        disp = self.verbose - 2 if self.verbose > 1 else -1
+        optimizer_params = {'method': 'L-BFGS-B',
+                            'fun': self._loss_grad_lbfgs,
+                            'jac': True,
+                            'args': (X_valid, y_valid, classes,
+                                     target_neighbors, grad_static,
+                                     use_sparse),
+                            'x0': transformation,
+                            'tol': self.tol,
+                            'options': dict(maxiter=self.max_iter, disp=disp),
+                            'callback': self._callback
+                            }
+
+        # Call the optimizer
+        self.n_iter_ = 0
+        opt_result = minimize(**optimizer_params)
+
+        # Reshape the solution found by the optimizer
+        self.components_ = opt_result.x.reshape(-1, X_valid.shape[1])
+
+        # Stop timer
+        t_train = time.time() - t_train
+        if self.verbose:
+            cls_name = self.__class__.__name__
+
+            # Warn the user if the algorithm did not converge
+            if not opt_result.success:
+                warn('[{}] LMNN did not converge: {}'.format(
+                    cls_name, opt_result.message),
+                     ConvergenceWarning)
+
+            print('[{}] Training took {:8.2f}s.'.format(cls_name, t_train))
+
+        # Optionally store information returned by the optimizer
+        if self.store_opt_result:
+            self.opt_result_ = opt_result
+
+        return self
+
+    def transform(self, X):
+        """Applies the learned transformation to the given data.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Data samples.
+
+        Returns
+        -------
+        X_embedded: array, shape (n_samples, n_components)
+            The data samples transformed.
+
+        Raises
+        ------
+        NotFittedError
+            If :meth:`fit` has not been called before.
+        """
+
+        check_is_fitted(self, ['components_'])
+        X = check_array(X)
+
+        return np.dot(X, self.components_.T)
```

Comment:
I'm not sure about this. Orthogonality is only enforced in the first iteration and only if PCA is used for initialization. Otherwise, the optimization is unconstrained.

Reply:
Exactly, this is why it may be helpful to future readers of that code, to leave a comment, why a simple dot product is necessary and sufficient for projection.

Reply:
Done.